temporal-sdk/src/test/java/io/temporal/workflow/LocalAsyncCompletionWorkflowTest.java

```diff
@@ -0,0 +1,145 @@
+import io.temporal.activity.ActivityExecutionContext;
+import io.temporal.activity.ActivityInterface;
+import io.temporal.activity.ActivityMethod;
+import io.temporal.activity.ActivityOptions;
+import io.temporal.client.ActivityCompletionClient;
+import io.temporal.client.WorkflowOptions;
+import io.temporal.common.RetryOptions;
+import io.temporal.worker.WorkerOptions;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+
+public class LocalAsyncCompletionWorkflowTest {
+
+  public static final int MAX_CONCURRENT_ACTIVITIES = 1;
+
+  @Rule
+  public TestWorkflowRule testWorkflowRule =
+      TestWorkflowRule.newBuilder()
+          .setWorkerOptions(
+              WorkerOptions.newBuilder()
+                  .setMaxConcurrentActivityExecutionSize(MAX_CONCURRENT_ACTIVITIES)
+                  .setActivityPollThreadCount(5)
+                  .build())
+          .setWorkflowTypes(TestWorkflowImpl.class)
+          .setActivityImplementations(new AsyncActivityWithManualCompletion())
+          .setUseExternalService(Boolean.parseBoolean(System.getenv("USE_DOCKER_SERVICE")))
+          .setTarget(System.getenv("TEMPORAL_SERVICE_ADDRESS"))
+          .build();
+
+  private static final AtomicInteger concurrentActivitiesCount = new AtomicInteger(0);
+
+  @WorkflowInterface
+  public interface TestWorkflow {
+
+    @WorkflowMethod
+    String execute(String taskQueue);
+  }
+
+  public static class TestWorkflowImpl implements TestWorkflow {
+
+    @Override
+    public String execute(String taskQueue) {
+      TestActivity activity =
+          Workflow.newActivityStub(
+              TestActivity.class,
+              ActivityOptions.newBuilder()
+                  .setScheduleToStartTimeout(Duration.ofSeconds(10))
+                  .setScheduleToCloseTimeout(Duration.ofSeconds(10))
+                  .setHeartbeatTimeout(Duration.ofSeconds(1))
+                  .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(1).build())
+                  .build());
+      List<Promise<Integer>> promises = new ArrayList<>();
+      for (int i = 0; i < 10; i++) {
+        promises.add(Async.function(() -> activity.execute(2)));
+      }
+      Promise.allOf(promises).get();
+      for (Promise<Integer> promise : promises) {
+        if (promise.getFailure() != null) {
+          return "exception";
+        }
+        if (promise.get() != 4) { // All activities compute 2 * 2
+          return "wrong result";
+        }
+      }
+      return "success";
+    }
+  }
+
+  @ActivityInterface
+  public interface TestActivity {
+
+    @ActivityMethod
+    int execute(int value);
+  }
+
+  public static class AsyncActivityWithManualCompletion implements TestActivity {
+
+    @Override
+    public int execute(int value) {
+      int concurrentActivities = concurrentActivitiesCount.incrementAndGet();
```

Comment:
This doesn't need to be static. Move to the AsyncActivityWithManualCompletion field.

Reply:
Done